app/src/main/java/com/epfl/drawyourpath/utils/Utils.kt

```diff
@@ -264,4 +269,91 @@ object Utils {
+        } catch (e: MlKitException) {
+            result.completeExceptionally(Error("The ML model identifier language tag failed to parse."))
+            return result
+        }
+        if (modelIdentifier == null) {
+            result.completeExceptionally(Error("The ML model was not found."))
+            return result
+        }
+
+        var model = DigitalInkRecognitionModel.builder(modelIdentifier!!).build()
+        val remoteModelManager = RemoteModelManager.getInstance()
+        remoteModelManager.isModelDownloaded(model)
+            .addOnSuccessListener() { is_downloaded: Boolean ->
+                if (!is_downloaded) {
+                    remoteModelManager.download(model, DownloadConditions.Builder().build())
+                        .addOnSuccessListener {
+                            Log.i("Utils", "Model downloaded")
+                            result.complete(model)
+                        }
+                        .addOnFailureListener { e: Exception ->
+                            result.completeExceptionally(Error("Fail to download the model."))
+                        }
+                } else {
+                    result.complete(model)
+                }
+            }
+            .addOnFailureListener { e: Exception ->
+                result.completeExceptionally(Error("The remote model manager could not verify if the model was downloaded."))
+            }
+        return result
+    }
+
+    /**
+     * Helper function to evaluate the drawing given the model
+     * @param ink the drawing as an Ink
+     * @param model the ML model
+     * @return the result of the ML model as a future of MLDrawingResults, null if it could not classify
+     */
+    fun recognizeDrawingML(ink: Ink, model: DigitalInkRecognitionModel): CompletableFuture<MLDrawingClassification> {
+        val recognizer: DigitalInkRecognizer =
+            DigitalInkRecognition.getClient(DigitalInkRecognizerOptions.builder(model).build())
+        var classificationML = CompletableFuture<MLDrawingClassification>()
+        recognizer.recognize(ink)
+            .addOnSuccessListener { result: RecognitionResult ->
+                if (result.candidates.isNotEmpty()) {
+                    // TODO : improve the custom score for display
+                    // The raw score can be positive or negative, and the lower the better
+                    var customScore = result.candidates[0].score!!
+                    classificationML.complete(
+                        MLDrawingClassification(result.candidates[0].text, result.candidates[0].score!!, customScore)
+                    )
+                } else {
+                    classificationML.complete(null)
+                }
+            }
+            .addOnFailureListener { e: Exception ->
+                classificationML.completeExceptionally(Error("Error while recognizing the model."))
```

Comment:
Can be helpful to return a message with the actual error message of the exception.

```suggestion
                classificationML.completeExceptionally(Error("Error while recognizing the model: ${e.localizedMessage}"))
```

Reply:
Yes that is true, I forgot to add the actual error message from the code itself.